### ISSUE TITLE
作者您好，有个关于参数的问题想要请教一下。

### ISSUE THREAD

**Issue by xiaozeng107** [DevilExileSu/transformer#3]
我与另一位提问者遇到了相似的问题，在运行“python test.py --sent 这是一个例子”时，
我得到的结果为：
['', '这是', '一个', '例子', '']
['', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the', 'the']

我采用的参数为  python train.py --batch_size 32 --h_dim 256 --lr 0.0005 --epochs 10
请问作者是否有其他推荐的参数，您训练时的参数是多少呢？

**Comment by DevilExileSu** [comment]
不好意思，这个项目之前的原始代码丢失了，第一次提交的代码不是该项目的最新代码，也就是README中例子所使用的模型的训练代码😂。当时也没有在意，后面也是在其他人的帮助下发现了代码不能运行的问题，后来进行了修改。现在我提交的时有进行了验证过的代码，可以尝试一下，参数就是README中的参数，但是收敛速度比较慢，可以修改下lr试试，我在验证时是0.0005和0.005交替着来的，中断然后修改参数 --resume true 这样再从断点开始使用新的lr进行训练，详细可以查看日志。
然后推荐你可以看看这个 https://github.com/bentrevett/pytorch-seq2seq 我当时参考了很多。

**Comment by xiaozeng107** [comment]
非常感谢您的回复！能否再请教一下？

我已经更新了代码，采用了参数python train.py --batch_size 32 --h_dim 256 --lr 0.0005 --epochs 50进行训练，在训练时遇到了这样的问题，或许是early stopping的原因训练提前终止：

2022-09-07 10:16:48,879 [INFO]: Train Epoch: 11, 5220/5292 (99%), Loss: 6.685256
2022-09-07 10:16:49,541 [INFO]: Train Epoch: 11, 5240/5292 (99%), Loss: 6.629090
2022-09-07 10:16:50,207 [INFO]: Train Epoch: 11, 5260/5292 (99%), Loss: 6.593773
2022-09-07 10:16:50,903 [INFO]: Train Epoch: 11, 5280/5292 (100%), Loss: 6.449338
2022-09-07 10:16:51,244 [INFO]: Train Epoch: 11, total Loss: 35001.308530, mean Loss: 6.614004
2022-09-07 10:16:51,244 [DEBUG]: start validation
2022-09-07 10:16:52,850 [INFO]: Train Epoch: 11, validation loss is : 7.480
2022-09-07 10:16:52,851 [DEBUG]: EarlyStopping counter:10 out of 10
2022-09-07 10:16:52,851 [DEBUG]: Training early stops

紧接着我运行test.py,得到了这样的结果：

2022-09-07 10:18:08,586 [DEBUG]: Config loaded from file config.json
这是一个例子
['<sos>', '这是', '一个', '例子', '<eos>']
['<sos>', '<eos>']

请问这种情况是什么原因造成的，或许有什么解决办法吗？

**Comment by xiaozeng107** [comment]
非常感谢您的回复！能否再请教一下？

我已经更新了代码，采用了参数python train.py --batch_size 32 --h_dim 256 --lr 0.0005 --epochs 50进行训练，在训练时遇到了这样的问题，或许是early stopping的原因训练提前终止：

2022-09-07 10:16:48,879 [INFO]: Train Epoch: 11, 5220/5292 (99%), Loss: 6.685256
2022-09-07 10:16:49,541 [INFO]: Train Epoch: 11, 5240/5292 (99%), Loss: 6.629090
2022-09-07 10:16:50,207 [INFO]: Train Epoch: 11, 5260/5292 (99%), Loss: 6.593773
2022-09-07 10:16:50,903 [INFO]: Train Epoch: 11, 5280/5292 (100%), Loss: 6.449338
2022-09-07 10:16:51,244 [INFO]: Train Epoch: 11, total Loss: 35001.308530, mean Loss: 6.614004
2022-09-07 10:16:51,244 [DEBUG]: start validation
2022-09-07 10:16:52,850 [INFO]: Train Epoch: 11, validation loss is : 7.480
2022-09-07 10:16:52,851 [DEBUG]: EarlyStopping counter:10 out of 10
2022-09-07 10:16:52,851 [DEBUG]: Training early stops

紧接着我运行test.py,得到了这样的结果：

2022-09-07 10:18:08,586 [DEBUG]: Config loaded from file config.json
这是一个例子
['<sos>', '这是', '一个', '例子', '<eos>']
['<sos>', '<eos>']

请问这种情况是什么原因造成的，或许有什么解决办法吗？

**Comment by xiaozeng107** [comment]
作者您好，对于上个问题，我有一些补充描述。
我们在训练时，大概训练两轮或三轮时便会出现：
[DEBUG]: EarlyStopping counter:1 out of 10

如果我们接着训练，随着train epoch的增加EarlyStopping counter的值也会增加，直至EarlyStopping counter=10便会自动停止训练。但我翻看了您上传的日志记录，似乎并没有出现这种情况，请问您是否知道这应该怎么解决？非常期盼您的回复！

**Comment by DevilExileSu** [comment]
在更新代码之前我也经常遇到这样的问题，不过后来更新代码在`encoder`和`decoder`中引入了`self.scale = torch.sqrt(torch.FloatTensor([h_dim])).to(device)`后，是没有再遇到这个问题了。在前面的参考链接里有提到，如果不加这个`scale`是很难进行训练的。
可以试着把`--early_stop`加大，我刚开始训练时使用`lr=0.005`，我的训练环境是`python=3.8`，`cuda=10.1`，`cudnn=7.6.5`， `pytorch-gpu=1.7.1`， 希望可以帮到你。

**Comment by xiaozeng107** [comment]
非常抱歉再打扰您，您的回复给了我们很多启发，但我们修改了很多参数，尝试了很多办法依旧无法解决，我们确认使用了已经更新的代码，encoder和decoder中也引入了self.scale =torch.sqrt(torch.FloatTensor([h_dim])).to(device)。

我们训练时，loss值始终是在第二次训练结束时最小，第二轮validation loss值为7.086，接着训练时的loss值一直没有下降，导致EarlyStopping counter一直增加直至停止训练。

请问您是否有已经训练完成的模型，能否上传以供我们研究学习？非常感谢您！！

**Comment by DevilExileSu** [comment]
链接：https://pan.baidu.com/s/1XcuE0fBz27NS0VkVK4kspw 
提取码：7vo3
这个就是README例子中使用的模型，也是日志中训练的模型，直接替换`saved_models/`中的`model_best.pt`，希望可以帮到你。
